python/pyarrow/_dataset.pyx

```diff
@@ -1998,6 +1998,41 @@ cdef class PartitioningFactory(_Weakrefable):
     cdef inline shared_ptr[CPartitioningFactory] unwrap(self):
         return self.wrapped
 
+    @property
+    def type_name(self):
+        return frombytes(self.factory.type_name())
+
+    def create_with_schema(self, schema):
```

Comment:
IMO it is fine that a Partitioning requires a schema to be constructed, at the low-level (C++/cython). Allowing users to pass just field names and not a full-blown Partitioning object from a schema, is something that we can also easily do in the high-level Python bindings, without needing extensive changes to the C++ code (https://github.com/apache/arrow/pull/11014)

Reply:
And note that for *reading*, we already do this: as a user, you in theory never have to use a PartitioningFactory, as you can either pass `partitioning="hive"` or `partitioning=["field", "names"]` (for directory partitioning) to the `dataset(..)` function. Those are both shortcuts for using `ds.partitioning()` in the case it returns a factory. 
But yes, we designed this API when there was only reading and not yet writing ..

Reply:
I didn't quite understand what you were doing earlier.  Now that I understand I withdraw my comment although I'm happy to proceed with the C++ changes (Antoine's suggestion makes them much less extensive) if it would help.

Maybe it would be simpler to just allow `write_dataset` to accept a "list of column names + partitioning format" or a partitioning object?  You could describe the `Partitioning` object as something that is used to represent the inferred partitioning created by dataset discovery mechanisms.

Reply:
> Maybe it would be simpler to just allow write_dataset to accept a "list of column names + partitioning format" or a partitioning object?

I am personally +1 on this (that's what I also mentioned on Zulip, and then we don't need to "misuse" the factories to pass along the field names). 
The main downside is that this requires another keyword to specify the type of partitioning, though, if we want to support hive-style this way (R solves that by having an extra keyword `hive_style`, https://arrow.apache.org/docs/r/reference/write_dataset.html, we could have a `partitioning_flavor="hive"`, although that might get a bit long). Not fully sure what I would prefer in the end.

Reply:
I changed `write_dataset` to accept `partitioning + partitioning_flavor`, see 
```
ds.write_dataset(table, tempdir, format='parquet',
                     partitioning=["b"], partitioning_flavor="hive")
```
from the test

So we are not using a factory anymore. I'll update the documentation if we are ok with this as the final api we want users to rely on.